src/lib.rs

```diff
@@ -38,354 +43,137 @@ pub struct LongLivedToken {
+            .send()
+            .await?;
 
-    pub fn refresh_token(&mut self) -> Result<(), errors::Error> {
-        let token = self.token.clone(); // This is dump but I have to do it apparently
-        let refresh_token: String;
-        match token {
-            Some(token) => {
-                refresh_token = match token {
-                    Token::Oauth(oauth) => oauth.refresh_token,
-                    Token::LongLived(_) => {
-                        return Err(errors::Error::Refresh());
-                    }
-                };
-            },
-            None => return Err(errors::Error::NoAuth())
-        }
-        
+        let api_states = resp.json::<Vec<HaEntityState>>().await?;
+        Ok(api_states)
+    }
 
-        let response = self
+    pub async fn register_machine(
+        &self,
+        request: &RegisterDeviceRequest,
+    ) -> Result<RegisterDeviceResponse, errors::Error> {
+        let endpoint = format!(
+            "http://{}/api/mobile_app/registrations",
+            self.instance_urls[0]
+        );
+        let token = self.get_token()?;
+        let resp = self
             .client
-            .post(&self.instance_urls[0])
-            .query(&[
-                ("grant_type", "refresh_token"),
-                ("client_id", CLIENT_ID),
-                ("refresh_token", refresh_token.as_str()),
-            ])
-            .send();
-        match executor::block_on(response) {
-            Ok(response) => {
-                match response.error_for_status() {
-                    Ok(response) => {
-                        match executor::block_on(response.json::<types::RefreshToken>()) {
-                            Ok(response_data) => {
-                                let oauth = OAuthToken {
-                                    token: response_data.access_token,
-                                    token_expiration: time::SystemTime::now()
-                                        + time::Duration::from_secs(response_data.expires_in),
-                                    refresh_token: refresh_token.to_string(),
-                                };
-                                self.token = Some(Token::Oauth(oauth));
-                                return Ok(());
-                            }
-                            Err(err) => {
-                                return Err(errors::Error::from(err));
-                            }
-                        };
-                    }
-                    Err(err) => {
-                        return Err(errors::Error::from(err));
-                    }
-                };
-            }
-            Err(err) => {
-                return Err(errors::Error::from(err));
-            }
-        };
+            .post(endpoint.as_str())
+            .header("Authorization", format!("Bearer {}", token))
+            .json(&request)
+            .send()
+            .await?;
+
+        let r: RegisterDeviceResponse = resp.json().await?;
```

Comment:
Not saving returned information

Reply:
Not quite sure I understand. What should we be saving here?

Reply:
we should not not requiring that the user wire together the webhook id etc.

Reply:
Sorry I still don't think I understand what you are asking. What do you mean wire together the webhook id? AFAIK the webhook is will always come back as part of this response.

Reply:
we should be saving the webhook id and not making the user do that. I'm fine with returning it so that they can access it but we should save it in the object.

Reply:
Actually never mind its goo that should be part of a bigger change to the API. That we can work on later.